diff --git a/qtbase/src/android/accessibility/jar/src/org/qtproject/qt5/android/accessibility/QtAccessibilityDelegate.java b/qtbase/src/android/accessibility/jar/src/org/qtproject/qt5/android/accessibility/QtAccessibilityDelegate.java
index cc2aba7..1abd05f 100644
--- a/qtbase/src/android/accessibility/jar/src/org/qtproject/qt5/android/accessibility/QtAccessibilityDelegate.java
+++ b/qtbase/src/android/accessibility/jar/src/org/qtproject/qt5/android/accessibility/QtAccessibilityDelegate.java
@@ -310,6 +310,9 @@ public class QtAccessibilityDelegate extends View.AccessibilityDelegate
 
         node.setClassName(m_view.getClass().getName() + DEFAULT_CLASS_NAME);
         node.setPackageName(m_view.getContext().getPackageName());
+	String id = QtNativeAccessibility.viewIdResourceForAccessibleObject(virtualViewId);
+	if (!TextUtils.isEmpty(id))
+            node.setViewIdResourceName(String.format("%s:id/%s", node.getPackageName(), id));
 
         if (!QtNativeAccessibility.populateNode(virtualViewId, node))
             return node;
diff --git a/qtbase/src/android/accessibility/jar/src/org/qtproject/qt5/android/accessibility/QtNativeAccessibility.java b/qtbase/src/android/accessibility/jar/src/org/qtproject/qt5/android/accessibility/QtNativeAccessibility.java
index f0bf4a0..cf31498 100644
--- a/qtbase/src/android/accessibility/jar/src/org/qtproject/qt5/android/accessibility/QtNativeAccessibility.java
+++ b/qtbase/src/android/accessibility/jar/src/org/qtproject/qt5/android/accessibility/QtNativeAccessibility.java
@@ -55,6 +55,7 @@ class QtNativeAccessibility
     static native boolean clickAction(int objectId);
     static native boolean scrollForward(int objectId);
     static native boolean scrollBackward(int objectId);
+    static native String viewIdResourceForAccessibleObject(int objectId);
 
     static native boolean populateNode(int objectId, AccessibilityNodeInfo node);
 }
diff --git a/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp b/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index 5c927da..433c8a3 100644
--- a/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ b/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -64,6 +64,7 @@ namespace QtAndroidAccessibility
     static jmethodID m_setScrollableMethodID = 0;
     static jmethodID m_setTextSelectionMethodID = 0;
     static jmethodID m_setVisibleToUserMethodID = 0;
+    static jmethodID m_setViewIdResourceNameMethodID = 0;
 
 
     static void setActive(JNIEnv */*env*/, jobject /*thiz*/, jboolean active)
@@ -213,6 +214,27 @@ if (!clazz) { \
         return descriptionForAccessibleObject_helper(env, iface);
     }
 
+    static jstring viewIdResourceForAccessibleObject_helper(JNIEnv *env, QAccessibleInterface *iface)
+    {
+        QString resourceId = "";
+        if (iface && iface->isValid()) {
+            QObject *object = iface->object();
+
+            resourceId = object->objectName();
+
+	    if (resourceId.isEmpty()) {
+                resourceId = object->property("id").toString();
+	    }
+        }
+        return env->NewString((jchar*) resourceId.constData(), (jsize) resourceId.size());
+    }
+
+    static jstring viewIdResourceForAccessibleObject(JNIEnv *env, jobject /*thiz*/, jint objectId)
+    {
+        QAccessibleInterface *iface = interfaceFromId(objectId);
+        return viewIdResourceForAccessibleObject_helper(env, iface);
+    }
+
     static bool populateNode(JNIEnv *env, jobject /*thiz*/, jint objectId, jobject node)
     {
         QAccessibleInterface *iface = interfaceFromId(objectId);
@@ -278,6 +300,7 @@ if (!clazz) { \
         {"clickAction", "(I)Z", (void*)clickAction},
         {"scrollForward", "(I)Z", (void*)scrollForward},
         {"scrollBackward", "(I)Z", (void*)scrollBackward},
+        {"viewIdResourceForAccessibleObject", "(I)Ljava/lang/String;", (jstring)viewIdResourceForAccessibleObject},
     };
 
 #define GET_AND_CHECK_STATIC_METHOD(VAR, CLASS, METHOD_NAME, METHOD_SIGNATURE) \
@@ -312,6 +335,7 @@ if (!clazz) { \
         GET_AND_CHECK_STATIC_METHOD(m_setFocusedMethodID, nodeInfoClass, "setFocused", "(Z)V");
         GET_AND_CHECK_STATIC_METHOD(m_setScrollableMethodID, nodeInfoClass, "setScrollable", "(Z)V");
         GET_AND_CHECK_STATIC_METHOD(m_setVisibleToUserMethodID, nodeInfoClass, "setVisibleToUser", "(Z)V");
+        GET_AND_CHECK_STATIC_METHOD(m_setViewIdResourceNameMethodID, nodeInfoClass, "setViewIdResourceName", "(Ljava/lang/String;)V");
 
         if (QtAndroidPrivate::androidSdkVersion() >= 18) {
             GET_AND_CHECK_STATIC_METHOD(m_setTextSelectionMethodID, nodeInfoClass, "setTextSelection", "(II)V");
diff --git a/qtdeclarative/src/quick/items/qquickitem.cpp b/qtdeclarative/src/quick/items/qquickitem.cpp
index 2722f48..eb68380 100644
--- a/qtdeclarative/src/quick/items/qquickitem.cpp
+++ b/qtdeclarative/src/quick/items/qquickitem.cpp
@@ -3425,6 +3425,18 @@ QQmlListProperty<QObject> QQuickItemPrivate::data()
                                              QQuickItemPrivate::data_clear);
 }
 
+QString QQuickItem::id() const
+{
+    QQuickItem *item = const_cast<QQuickItem *>(this);
+    QQmlEngine *engine = qmlEngine(item);
+    if (engine) {
+        QQmlContext *context = engine->contextForObject(item);
+        if (context)
+            return context->nameForObject(item);
+    }
+    return QString();
+}
+
 /*!
     \qmlpropertygroup QtQuick::Item::childrenRect
     \qmlproperty real QtQuick::Item::childrenRect.x
diff --git a/qtdeclarative/src/quick/items/qquickitem.h b/qtdeclarative/src/quick/items/qquickitem.h
index c17f6a6..62783da 100644
--- a/qtdeclarative/src/quick/items/qquickitem.h
+++ b/qtdeclarative/src/quick/items/qquickitem.h
@@ -97,6 +97,8 @@ class Q_QUICK_EXPORT QQuickItem : public QObject, public QQmlParserStatus
     Q_PRIVATE_PROPERTY(QQuickItem::d_func(), QQmlListProperty<QObject> resources READ resources DESIGNABLE false)
     Q_PRIVATE_PROPERTY(QQuickItem::d_func(), QQmlListProperty<QQuickItem> children READ children NOTIFY childrenChanged DESIGNABLE false)
 
+    Q_PROPERTY(QString id READ id)
+
     Q_PROPERTY(qreal x READ x WRITE setX NOTIFY xChanged FINAL)
     Q_PROPERTY(qreal y READ y WRITE setY NOTIFY yChanged FINAL)
     Q_PROPERTY(qreal z READ z WRITE setZ NOTIFY zChanged FINAL)
@@ -200,6 +202,7 @@ public:
     QRectF childrenRect();
     QList<QQuickItem *> childItems() const;
 
+    QString id() const;
     bool clip() const;
     void setClip(bool);
 
